File: analysis/tools/state_extract.py
#!/usr/bin/env python3
"""state_extract.py -- extract iwram.bin/ewram.bin from an mGBA GBA savestate.

Supported containers (format probed empirically against a simultaneous RAM dump,
see analysis/dumps/statepair/):

- PNG savestate (mGBA default, incl. GUI .ss0-.ss9 slot files): a PNG whose
  `gbAs` chunk is the zlib-compressed 0x61000-byte serialized state.
- Raw serialized state (0x61000 bytes, e.g. scripting saveStateFile without the
  screenshot flag).
- libretro mGBA-core .st* (e.g. MinUI handhelds): the 0x61000-byte serialized
  state followed by appended savedata/extra of core-dependent length. Detected
  by the state's version magic (u32 0x010000xx at +0) or the ROM title at +0x10;
  only the first 0x61000 bytes are used.

Within the 0x61000-byte GBA serialized state: IWRAM (0x8000 bytes) at +0x19000,
EWRAM (0x40000 bytes) at +0x21000 (verified byte-identical to a same-frame
emu:readRange dump).

Usage:
    python3 state_extract.py <savestate> [-o OUTDIR]

Writes OUTDIR/iwram.bin and OUTDIR/ewram.bin (default: alongside the state file,
in <state-basename>.ramdump/), then parse with:
    python3 parse_ram.py OUTDIR

Flash .sav files are NOT savestates and are out of scope.
"""

import argparse
import os
import struct
import sys
import zlib

STATE_SIZE = 0x61000
IWRAM_OFF, IWRAM_SIZE = 0x19000, 0x8000
EWRAM_OFF, EWRAM_SIZE = 0x21000, 0x40000
PNG_MAGIC = bytes.fromhex("89504e470d0a1a0a")


def deserialize(blob):
    """Return the raw 0x61000-byte serialized state from a savestate file blob."""
    if blob[:8] == PNG_MAGIC:
        off = 8
        while off + 12 <= len(blob):
            (length,) = struct.unpack_from(">I", blob, off)
            ctype = blob[off + 4:off + 8]
            if ctype == b"gbAs":
                state = zlib.decompress(blob[off + 8:off + 8 + length])
                if len(state) != STATE_SIZE:
                    raise ValueError(
                        "gbAs chunk decompressed to %#x bytes (expected %#x) -- "
                        "not a GBA savestate?" % (len(state), STATE_SIZE))
                return state
            off += 12 + length
        raise ValueError("PNG file has no gbAs chunk -- not an mGBA savestate")
    if len(blob) == STATE_SIZE:
        return blob
    if len(blob) > STATE_SIZE:
        # libretro-style container: serialized state first, appended extras after.
        # Verify it actually starts with a GBA state (version magic 0x010000xx, or
        # the 12-byte ROM title at +0x10) rather than blindly slicing.
        (magic,) = struct.unpack_from("<I", blob, 0)
        if (magic & 0xFFFF0000) == 0x01000000 or blob[0x10:0x1C] == b"POKEMON FIRE":
            return blob[:STATE_SIZE]
        raise ValueError(
            "file is larger than a GBA state (%#x > %#x) but does not start "
            "with an mGBA state header" % (len(blob), STATE_SIZE))
    raise ValueError(
        "unrecognized savestate: not a PNG and smaller than %#x bytes (got %#x). "
        "Note: flash .sav files are not savestates." % (STATE_SIZE, len(blob)))


def extract(state_path, out_dir=None):
    with open(state_path, "rb") as f:
        blob = f.read()
    state = deserialize(blob)
    if out_dir is None:
        base = os.path.splitext(os.path.basename(state_path))[0]
        out_dir = os.path.join(os.path.dirname(os.path.abspath(state_path)),
                               base + ".ramdump")
    os.makedirs(out_dir, exist_ok=True)
    iwram = state[IWRAM_OFF:IWRAM_OFF + IWRAM_SIZE]
    ewram = state[EWRAM_OFF:EWRAM_OFF + EWRAM_SIZE]
    if not any(iwram):
        sys.stderr.write("warning: IWRAM is all zero -- savestate may be invalid "
                         "or from a just-reset system\n")
    with open(os.path.join(out_dir, "iwram.bin"), "wb") as f:
        f.write(iwram)
    with open(os.path.join(out_dir, "ewram.bin"), "wb") as f:
        f.write(ewram)
    return out_dir


def main():
    ap = argparse.ArgumentParser(
        description="Extract iwram.bin/ewram.bin from an mGBA GBA savestate.")
    ap.add_argument("state", help="savestate file (PNG or raw 0x61000-byte state)")
    ap.add_argument("-o", "--out", help="output directory (default: <state>.ramdump/)")
    args = ap.parse_args()
    out_dir = extract(args.state, args.out)
    print(out_dir)


if __name__ == "__main__":
    main()

File: lib/parser/state-extract.js
/**
 * state-extract.js -- extract IWRAM/EWRAM from an mGBA GBA savestate.
 *
 * JS port of research/tools/state_extract.py. Supported containers:
 *
 * - PNG savestate (mGBA default, incl. GUI .ss0-.ss9 slot files): a PNG whose
 *   `gbAs` chunk is the zlib-compressed 0x61000-byte serialized state.
 * - Raw serialized state (0x61000 bytes).
 * - libretro mGBA-core .st* (MinUI handhelds): the 0x61000-byte serialized state
 *   followed by appended savedata of core-dependent length.
 *
 * Within the 0x61000-byte state: IWRAM (0x8000) at +0x19000, EWRAM (0x40000) at
 * +0x21000.
 *
 * Pure ES module: no Node/Bun APIs. Inflate uses DecompressionStream, so the API
 * is async.
 */

export const STATE_SIZE = 0x61000;
export const IWRAM_OFF = 0x19000;
export const IWRAM_SIZE = 0x8000;
export const EWRAM_OFF = 0x21000;
export const EWRAM_SIZE = 0x40000;

const PNG_MAGIC = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];

/** Error for an unrecognized/unusable savestate container. */
export class StateError extends Error {
  constructor(message) {
    super(message);
    this.name = "StateError";
  }
}

/** Python's "%#x": lowercase hex with an 0x prefix (0 renders as "0x0"). */
function hashHex(n) {
  return (n < 0 ? "-0x" : "0x") + Math.abs(n).toString(16);
}

function startsWithPng(blob) {
  if (blob.length < 8) return false;
  return PNG_MAGIC.every((b, i) => blob[i] === b);
}

async function inflate(bytes) {
  const stream = new Blob([bytes]).stream().pipeThrough(new DecompressionStream("deflate"));
  const chunks = [];
  let total = 0;
  const reader = stream.getReader();
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    chunks.push(value);
    total += value.length;
  }
  const out = new Uint8Array(total);
  let at = 0;
  for (const c of chunks) {
    out.set(c, at);
    at += c.length;
  }
  return out;
}

/**
 * Return the raw 0x61000-byte serialized state from a savestate file blob.
 * @param {Uint8Array} blob
 * @returns {Promise<Uint8Array>}
 */
export async function deserialize(blob) {
  if (startsWithPng(blob)) {
    const view = new DataView(blob.buffer, blob.byteOffset, blob.byteLength);
    let off = 8;
    while (off + 12 <= blob.length) {
      const length = view.getUint32(off, false); // PNG chunk lengths are big-endian
      const ctype = String.fromCharCode(blob[off + 4], blob[off + 5], blob[off + 6], blob[off + 7]);
      if (ctype === "gbAs") {
        const state = await inflate(blob.subarray(off + 8, off + 8 + length));
        if (state.length !== STATE_SIZE) {
          throw new StateError(
            `gbAs chunk decompressed to ${hashHex(state.length)} bytes (expected ` +
              `${hashHex(STATE_SIZE)}) -- not a GBA savestate?`,
          );
        }
        return state;
      }
      off += 12 + length;
    }
    throw new StateError("PNG file has no gbAs chunk -- not an mGBA savestate");
  }
  if (blob.length === STATE_SIZE) return blob;
  if (blob.length > STATE_SIZE) {
    // libretro-style container: serialized state first, appended extras after.
    // Verify it actually starts with a GBA state (version magic 0x010000xx, or the
    // 12-byte ROM title at +0x10) rather than blindly slicing.
    const magic = new DataView(blob.buffer, blob.byteOffset, blob.byteLength).getUint32(0, true);
    const title = String.fromCharCode(...blob.subarray(0x10, 0x1c));
    if ((magic & 0xffff0000) >>> 0 === 0x01000000 || title === "POKEMON FIRE") {
      return blob.subarray(0, STATE_SIZE);
    }
    throw new StateError(
      `file is larger than a GBA state (${hashHex(blob.length)} > ${hashHex(STATE_SIZE)}) ` +
        "but does not start with an mGBA state header",
    );
  }
  throw new StateError(
    `unrecognized savestate: not a PNG and smaller than ${hashHex(STATE_SIZE)} bytes ` +
      `(got ${hashHex(blob.length)}). Note: flash .sav files are not savestates.`,
  );
}

/**
 * Extract the two RAM regions from a savestate file blob.
 * @param {Uint8Array} blob raw bytes of a savestate file
 * @returns {Promise<{iwram: Uint8Array, ewram: Uint8Array}>}
 */
export async function extractRam(blob) {
  const state = await deserialize(blob);
  return {
    iwram: state.subarray(IWRAM_OFF, IWRAM_OFF + IWRAM_SIZE),
    ewram: state.subarray(EWRAM_OFF, EWRAM_OFF + EWRAM_SIZE),
  };
}
